Star rating: scale with library font

ronso0 opened the pull request:
Helpful with font sizes < ~70% or > ~150% of the default font size.
Ratings in decks and Track Info dialog still use the default scale factor.

![image](https://github.com/mixxxdj/mixxx/assets/5934199/bde9c195-8c7c-467f-bc7c-0c98f0dbe897)
![image](https://github.com/mixxxdj/mixxx/assets/5934199/bc46818f-9049-4007-b445-b7bab2ed750e)

Previous size with default library font:
![image](https://github.com/mixxxdj/mixxx/assets/5934199/0e280be4-3580-4836-bebf-9972962e3c03)


Played checkbox is next ;)

ronso0 committed:
Star rating: reset if pointer leaves rect at the right side

ronso0 committed:
Star rating: sizeHint uses scalefactor, for auto-adjusting table column

daschuer commented on src/library/starrating.cpp:
```diff
@@ -30,20 +31,21 @@ StarRating::StarRating(
     painter->setPen(Qt::NoPen);
 
-    int yOffset = (rect.height() - PaintingScaleFactor) / 2;
+    int yOffset = (rect.height() - m_scaleFactor) / 2;
```
subtracting a factor form a value sound wrong. If this is relay correct here, please add am comment.

daschuer reviewed (changes requested):
some first comments. Not yet tested.

ronso0 committed:
Star rating: scale with library font

ronso0 commented:
I rebased onto 2.4 (removed the unrelated BPM changes and added a comment to the paint method)

daschuer commented:
This branch creates way smaller stars. Is that intended? I do not demand the old size. 
New:
![grafik](https://github.com/mixxxdj/mixxx/assets/1777442/d227a67e-4143-4065-92d7-54dd684ab562)
Old:
![grafik](https://github.com/mixxxdj/mixxx/assets/1777442/1bae09b3-36c2-41e8-a334-38629cab64df)
The scaling by font works.

ronso0 commented on src/library/starrating.cpp:
```diff
@@ -30,20 +31,22 @@ StarRating::StarRating(
 
     //determine number of stars that are possible to paint
-    int n = rect.width() / PaintingScaleFactor;
+    int n = rect.width() / m_scaleFactor;
```
`mixxx::TrackRecord::kMaxRating` from https://github.com/mixxxdj/mixxx/blob/b4a40b876cfeafd6b5f3926a8468d6a2e6cb40f4/src/track/trackrecord.h#L78

ronso0 replied:
`m_maxStarCount` is the paint limit, the painter only draws polygons (star or diamond) up to that number.

There's no limit for the rect, sure it can be wider than 5 stars. The StarEditor considers all mouse events in the rect, see `StarEditor::starAtPosition()`.

ronso0 committed:
Star rating: improve comments

daschuer commented on src/library/starrating.h:
```diff
@@ -22,11 +22,15 @@ class StarRating {
 
     static constexpr int kMinStarCount = 0;
 
+    static QSize getBaseSize() {
```
From the function name it is not obvious what base size is. With all the HID scaling in mind this sounds form me as a normal and usable size. But it actually gives only the ratio between the width and length, normalized on a 1 pixel height widget. I have considered better names like `getMinimumSize()` or `getSizeForHeightOne()`, but then I realized that the scale factor is always the desired height and I have proposed to refactor it to `return getSizeForHight(int height);`
and rename `m_scaleFactor`  to `m_height`. 

If you like to keep the scale factor nomenclature, a better function name or/and explaining comment would be nice.

ronso0 replied:
Sorry for dropping the dsicussion. I considered the scale factor alternatives you proposed but none really fits IMHO. 'scale factor' is used only in this class so I'd like to keep that.

Regarding the get function, what about `getNormalizedSize()`? If you don't think that fits: for circles there's 'unit circle', so maybe we could call it `getUnitSize()`, though that might be understood as size of some self-contained element, not that of the entire stars container. Hence, I prefer `getNormalizedSize()`.
I'll also add a comment, of course.

ronso0 commented:
I'll put this on hold for now.

I'll try to adjust the size to match the default settings (library font type & size).

daschuer reviewed (commented):
Just found old not published review comments:

ronso0 commented on src/library/stardelegate.cpp:
```diff
@@ -27,14 +27,23 @@ void StarDelegate::paintItem(
+    // Make stars rectangle as tall as Capitals of current font.
+    // get capital height and create rect from that
+    QFontMetrics mf(option.font);
+    int height = mf.capHeight();
```
> Did you consider the smaller stars issue? 

What do you mean?

> You have currently picked the size of "H" and not "Â"
Was that intended?

Yes.

daschuer replied:
The issue shown in this post:
https://github.com/mixxxdj/mixxx/pull/11772#issuecomment-1667234240

What is your rational for using smaller stars? 
I am not against it. I just want to make sure it has been done on purpose.

daschuer replied:
Than just take 
   (ascent() + capHeight())/2

ronso0 replied:
yup, that's what I did, and it looks good.
will push it soon

ronso0 committed:
star widgets: move starAtPosition(int) to StarRating

ronso0 commented:
Alright, we now have a suitable star size (IMO), it's (capital height - ascent height) / 2.
I hope I addressed all review comments.

Also, I did some more cleanup and consolidation to keep maintenance low.
